### PR TITLE
Remove primitives

### PR DESCRIPTION
**Pull request by mbianco**
Removing threading primitives (and barrier) from GHEX. This simplifies the overall structure and eliminates the need for tokens, which are cumbersome and confusing for the user.

There is a semantic change, though. When `get_communicator()` is called (now without the token), a communicator is created and provided to the user. Before, `get_communicator` with the same token was giving back the same communicator, but now a new one is created every time.

Still, a context is generated by the factory, and given to the user, who has the responsibility of destroying it before `MPI_Finalize`. I created a couple os proposal (one using shared library - to be extended maybe to be more "erased"), and one extending the approach we already have - but still not the cleanest. You can find the code that makes `test_low_level` working at https://github.com/mbianco/GHEX/pull/2 and https://github.com/mbianco/GHEX/pull/3

### CODE REVIEW

**Comment by angainor** [comment]
@mbianco I see you used `std::mutex`. I will test the performance with this PR (is it ready to roll?), but from what I remember this was much slower than what we ended up using (a recursive pthread spinlock). Did we agree that we require pthread to be supported? can we abstract it out so that we can use pthread spin lock instead? and make it recursive..

File: cmake/FindUCP.cmake [hunk 1]
```diff
@@ -20,6 +20,10 @@ find_library(UCP_LIBRARY HINT ${UCP_DIR} NAMES ucp
 set(UCP_LIBRARIES    ${UCP_LIBRARY} CACHE INTERNAL "")
 set(UCP_INCLUDE_DIRS ${UCP_INCLUDE_DIR} CACHE INTERNAL "")
 
+message("CIAODIR ${UCX_DIR}")
+message("CIAO ${UCP_LIBRARY}")
+message("CIAO2 ${UCP_INCLUDE_DIR}")
+
```
Comment (mbianco) [inline cmake/FindUCP.cmake:26]:
```suggestion
```

File: benchmarks/transport/ghex_p2p_bi_cb_avail_mt.cpp [hunk 2]
```diff
@@ -151,7 +150,10 @@ int main(int argc, char *argv[])
                 make_zero(rmsgs[j]);
             }
 
-            comm.barrier();
+#ifdef USE_OPENMP
+#pragma omp barrier
+#endif
+            MPI_Barrier(MPI_COMM_WORLD);
```
Comment (angainor) [inline benchmarks/transport/ghex_p2p_bi_cb_avail_mt.cpp:156]: @mbianco We cannot execute MPI barrier here. It has to be our barrier, which would progress the communication engine. MPI barrier does not progress our workers and leads to a dead-lock.

The previous barrier implementation from `communicator` has to be available somehow and executed here.

File: include/ghex/transport_layer/ucx/communicator.hpp [hunk 3]
```diff
@@ -113,11 +112,9 @@ namespace gridtools {
                     template <typename Message>
                     [[nodiscard]] future<void> recv(Message &msg, rank_type src, tag_type tag)
                     {
-                        const auto rtag = ((std::uint_fast64_t)tag << 32) | 
+                        const auto rtag = ((std::uint_fast64_t)tag << 32) |
                                            (std::uint_fast64_t)(src);
-                        return m_send_worker->m_thread_primitives->critical(
-                            [this,rtag,&msg,src,tag]()
-                            {
+                        std::lock_guard<decltype(m_send_worker->mutex())> lock(m_send_worker->mutex());
```
Comment (boeschf) [inline include/ghex/transport_layer/ucx/communicator.hpp:117]: this will only work for the spin-lock recursive mutex. A standard mutex is not recursive and we'll need a `unique_lock` for that. If we stick to our custom mutex for ucx, then everything is fine.

File: include/ghex/transport_layer/ucx/communicator.hpp [hunk 4]
```diff
@@ -160,25 +155,22 @@ namespace gridtools {
                         gridtools::ghex::tl::cb::progress_status status;
                         int p = 0;
                         p+= ucp_worker_progress(m_ucp_sw);
-                        p+= ucp_worker_progress(m_ucp_sw);
-                        p+= ucp_worker_progress(m_ucp_sw);
+
```
[the remark this answers is not in the document]
Comment (boeschf) [reply include/ghex/transport_layer/ucx/communicator.hpp:158]: because it gave better performance - @angainor did do some more benchmarks and this now in conjunction with the sched_yield below should be ok, am I correct on this?

File: include/ghex/transport_layer/ucx/context.hpp [hunk 5]
```diff
@@ -166,25 +164,22 @@ namespace gridtools {
...
                 }
-                
+
                 communicator_type get_serial_communicator()
                 {
                     return {&m_worker,&m_worker};
                 }
 
-                communicator_type get_communicator(const thread_token& t)
+                communicator_type get_communicator()
```
Comment (boeschf) [inline include/ghex/transport_layer/ucx/context.hpp:178]: same here for the above serial communicator - I think it is not needed anymore

File: include/ghex/transport_layer/ucx/request.hpp [hunk 6]
```diff
@@ -100,27 +93,24 @@ namespace gridtools{
...
                     }
                 };
-                using request_data_size_cb = std::integral_constant<std::size_t,
-                sizeof(request_data_cb<::gridtools::ghex::threads::atomic::primitives>) +
-                alignof(request_data_cb<::gridtools::ghex::threads::atomic::primitives>)>;
 
                 /** @brief minimum size of the ucx provided storage in bytes */
-                using request_data_size = request_data_size_cb;
-                
+                using request_data_size = std::integral_constant<std::size_t,sizeof(request_data_cb)>;
```
Comment (boeschf) [inline include/ghex/transport_layer/ucx/request.hpp:104]: why did you remove the alignment padding? Not sure if this is safe since we have no idea what size the ucx will use for its own request.

File: include/ghex/transport_layer/util/barrier.hpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ */
+
+namespace gridtools {
+    namespace ghex {
+        namespace tl {
+
+            template <typename TLCommunicator>
+            void barrier(MPI_Comm comm, TLCommunicator& tlcomm)
```
[the remark this answers is not in the document]
Comment (angainor) [reply include/ghex/transport_layer/util/barrier.hpp:17]: @boeschf  Do you mean the MPI communicator?

File: include/ghex/transport_layer/context.hpp [hunk 8]
```diff
@@ -111,16 +100,9 @@ namespace gridtools {
...
                   * This function is thread-safe. */
-                communicator_type get_communicator(const thread_token& t)
-                {
-                    return m_transport_context.get_communicator(t);
-                }
-
-                /** @brief return a per-thread thread token.
-                  * This function is thread-safe. */
-                thread_token get_token() noexcept
+                communicator_type get_communicator(/*const thread_token& t*/)
```
[the remark this answers is not in the document]
Comment (angainor) [reply include/ghex/transport_layer/context.hpp:103]: @boeschf If I remember correctly, for UCX, the serial communicator would return a communicator where the send and recv workers would be one actual worker. That is not the same, and might still be interesting / needed.

File: include/ghex/transport_layer/util/barrier.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (boeschf) [reply include/ghex/transport_layer/util/barrier.hpp:17]: I mean that several threads can enter this function and then call the mpi barrier which may cause problems - actually, I am not sure what the semantics should be in this case.

File: include/ghex/transport_layer/context.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (boeschf) [reply include/ghex/transport_layer/context.hpp:103]: true, this is not the same, but is not used anywhere except in the setup phase which is not performance critical. I am just saying that a simpler interface would do just as fine for now.

File: include/ghex/transport_layer/ucx/communicator.hpp [hunk 9]
```diff
@@ -1,46 +1,45 @@
...
+ *
  */
 #ifndef INCLUDED_GHEX_TL_UCX_COMMUNICATOR_CONTEXT_HPP
 #define INCLUDED_GHEX_TL_UCX_COMMUNICATOR_CONTEXT_HPP
 
 #include <atomic>
+#include <mutex>
 #include "../shared_message_buffer.hpp"
 #include "./future.hpp"
+#include "../util/pthread_spin_mutex.hpp"
```
[the remark this answers is not in the document]
Comment (angainor) [reply include/ghex/transport_layer/ucx/communicator.hpp:18]: the `std::mutex` was really really slow with multi-threading. I don't see much to configure for here, unless we find a performant alternative to pthread spin lock.

File: include/ghex/transport_layer/ucx/communicator.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (angainor) [reply include/ghex/transport_layer/ucx/communicator.hpp:158]: The multi-progress version has some performance benefits on our Intel-based platform for up to ~8 threads. Now that I tested this on up to 64 threads per  rank I can see that 1) we have to have a `sched_yield` call and 2) we should not call UCX progress more than once. Otherwise I observed lots of performance 'lock-ups', probably because one (or few) threads would dominate the execution and not let others progress the communication.

File: include/ghex/transport_layer/util/barrier.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (angainor) [reply include/ghex/transport_layer/util/barrier.hpp:17]: This is defined by MPI. If the communicator is duplicated, multiple threads can call `MPI_Barrier`. Otherwise you get undefined behavior. We don't have to do anything about this - users must provide a correct MPI communicator here.

File: include/ghex/transport_layer/context.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (angainor) [reply include/ghex/transport_layer/context.hpp:103]: I guess it is not crucial to have, you're right. I used this for experimenting recently, but that I can do off-line anyway. Let's clean up the interface.

File: include/ghex/transport_layer/util/barrier.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (angainor) [reply include/ghex/transport_layer/util/barrier.hpp:17]: One important thing though is that we have to document that all threads MUST call this barrier. Otherwise UCX will not be progressed correctly. So this would be wrong
```
// UCX comm phase
#pramga omp master
barrier(...);
#pragma omp barrier
```
because some master threads could potentially not reach the barrier and hang on the UCX comm phase, while the peer threads would hang (without progress) on the omp barrier.

[the remark this answers is not in the document]
Comment (boeschf) [reply include/ghex/transport_layer/util/barrier.hpp:17]: The call to the barrier is thread-safe, yes. And yes, all threads need to call the barrier. But neither we nor MPI can infer how many threads are running (and whether the number of threads is even constant per rank). This is quite a difficult problem, I think, and would have been easily solved with our previous primitives. Right now, I don't see a good solution.

[the remark this answers is not in the document]
Comment (angainor) [reply include/ghex/transport_layer/util/barrier.hpp:17]: @boeschf I think I see what you mean. We have to keep progressing UCX until all threads have completed the barrier, not each individual thread. That's a problem. We could take an `nthreads` argument here, and keep calling UCX progress untill all threads have completed the MPI barrier. This would have the advantage that threads could pass `MPI_COMM_NULL`, in which case we would not require a special thread communicator. That would also solve the case with different number of threads per rank I think.

[the remark this answers is not in the document]
Comment (angainor) [reply include/ghex/transport_layer/util/barrier.hpp:17]: It seems we would only need an atomic variable to do this.

[the remark this answers is not in the document]
Comment (boeschf) [reply include/ghex/transport_layer/util/barrier.hpp:17]: I think the implementation will be more complex since you'll need to single out one thread (let's say thread 0) that does the Ibarrier (while still executing the progress in the wait) and the others just progress. There is no guarantee that all communications have finished one thread 0 completes the Ibarrier, however. It would only guarantee that all threads on one rank have finished communication with their thread 0 and that thread 0 has finished communicating with all threads on all ranks. All other communications may or may not be complete. If we only allow the same number of threads per rank then the problem may be simplified, but not by much. It would then work somehow like you describe above, but I think this would still not be enough as some communications may still be outstanding at this point. If a barrier is called for all possible thread-thread combinations (N-1)*N/2, then this may be enough... not sure though.

[the remark this answers is not in the document]
Comment (angainor) [reply include/ghex/transport_layer/util/barrier.hpp:17]: The procedure would be as follows:
- user calls `barrier` for each thread, passing an MPI communicator or `MPI_COMM_NULL`
- we call `MPI_Ibarrier` for all non-null communicators, and call our ucx progress in the test loop. User has to make sure that the MPI communicator has been constructed in a thread-aware manner (MPI standard)
- once out of the test loop, we increase the atomic counter (also for null communicators)
- then we get to the pure ucx progress loop, which we do until the atomic counter reaches the `nthreads` argument

Threads with NULL communicator would increase the counter immediately and proceed to the ucx progress loop.

We don't have to assure completion of UCX comm, we only have to assure that progress is called. This will assure that we leave the `barrier()` function. Then it's the user's responsibility.

File: include/ghex/transport_layer/ucx/context.hpp (hunk 5 above)
Comment (mbianco) [reply include/ghex/transport_layer/ucx/context.hpp:178]: I agree

File: include/ghex/transport_layer/ucx/request.hpp (hunk 6 above)
Comment (mbianco) [reply include/ghex/transport_layer/ucx/request.hpp:104]: They were not used anyway, IIRC

File: include/ghex/transport_layer/ucx/worker.hpp [hunk 10]
```diff
@@ -70,24 +70,22 @@ namespace gridtools {
...
                     address_t               m_address;
                     cache_type              m_endpoint_cache;
                     int                     m_progressed_sends = 0;
+                    mutex_t*                m_mutex_ptr;
                     volatile int            m_progressed_recvs = 0;
                     volatile int            m_progressed_cancels = 0;
 
                     worker_t() = default;
-                    worker_t(transport_context_type* c, thread_primitives_type* tp, thread_token* t, ucs_thread_mode_t mode);
+                    worker_t(transport_context_type* c, mutex_t& mm, ucs_thread_mode_t mode);
```
[the remark this answers is not in the document]
Comment (mbianco) [reply include/ghex/transport_layer/ucx/worker.hpp:88]: probably better, yes... Will fix it when I find why tests hang in containers

File: include/ghex/transport_layer/ucx/request.hpp (hunk 6 above)
Comment (boeschf) [reply include/ghex/transport_layer/ucx/request.hpp:104]: see https://github.com/GridTools/GHEX/blob/6c04d22c30a22956e7cc079819360e253535ac8f/include/ghex/transport_layer/ucx/context.hpp#L130

File: include/ghex/transport_layer/ucx/communicator.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (mbianco) [reply include/ghex/transport_layer/ucx/communicator.hpp:18]: I do not see why we should make the type parametric here...

File: include/ghex/transport_layer/ucx/communicator.hpp (hunk 3 above)
Comment (mbianco) [reply include/ghex/transport_layer/ucx/communicator.hpp:117]: I should just put the type instead of `decltype`, so if we change something we trigger an error and we check

Comment (boeschf) [reply include/ghex/transport_layer/ucx/communicator.hpp:117]: yes, sounds good

File: include/ghex/transport_layer/util/barrier.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (mbianco) [reply include/ghex/transport_layer/util/barrier.hpp:17]: MPI_Comm should not be an argument, only tl::comm, from which you get contexts, MPI_comm, etc.
What is missing is the number of threads that participate....

File: include/ghex/transport_layer/util/barrier.hpp [hunk 11]
```diff
@@ -70,7 +70,8 @@ namespace gridtools {
                 template <typename TLCommunicator>
                 void operator()(TLCommunicator& tlcomm) const
                 {
-                    in_node1_full(tlcomm);
+                    if (in_node1())
+                        rank_barrier(tlcomm);
```
Comment (boeschf) [inline include/ghex/transport_layer/util/barrier.hpp:74]: Not sure here: shouldn't the other threads also progress until the main thread (which does the mpi barrier) indicates that they can stop progressing?

Comment (angainor) [reply include/ghex/transport_layer/util/barrier.hpp:74]: @boeschf right. All while loops should call the progress function..

Comment (boeschf) [inline include/ghex/transport_layer/util/barrier.hpp:74]:
```suggestion
                    if (in_node1())
                        rank_barrier(tlcomm);
                    else
                    {
                        while(b_count2 == m_threads)
                            tlcomm.progress();
                    }
```

Comment (mbianco) [reply include/ghex/transport_layer/util/barrier.hpp:74]: @angainor when polling in the thread-barrier ucx gave me errors (access violation on very small addresses... not sure why). I assumed the thread barrier was fast enough. I can try when the tests pass reliably. Now they don't.

File: include/ghex/transport_layer/util/barrier.hpp [hunk 12]
```diff
@@ -128,7 +127,8 @@ namespace gridtools {
                 void in_node2() const
                 {
                     size_t ex = b_count2;
-                    while(!b_count2.compare_exchange_weak(ex, ex-1, std::memory_order_relaxed)) ex = b_count2;
+                    while(!b_count2.compare_exchange_weak(ex, ex-1, std::memory_order_relaxed))
+                        ex = b_count2;
                     if (ex == 1) {
```
Comment (angainor) [inline include/ghex/transport_layer/util/barrier.hpp:137]: @mbianco Should this not be `ex == 0` ?

Comment (angainor) [reply include/ghex/transport_layer/util/barrier.hpp:137]: Also, in `in_node1` it should read `expected == m_threads` instead of `expected == m_threads-1`

File: include/ghex/transport_layer/util/barrier.hpp [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+                    while(true) {
+                        tlcomm.progress();
+                        MPI_Test(&req, &flag, MPI_STATUS_IGNORE);
+                        if(flag) break;
+                    }
+                }
+
+                /**
+                 * This function synchronize the threads in a rank. The threads must obtain a token
+                 * before being able to call this function.
```
[the remark this answers is not in the document]
Comment (mbianco) [reply include/ghex/transport_layer/util/barrier.hpp:108]: leftover

File: include/ghex/transport_layer/context.hpp (hunk 8 above)
Comment (mbianco) [inline include/ghex/transport_layer/context.hpp:103]:
```suggestion
                communicator_type get_communicator()
```

Comment (mbianco) [inline include/ghex/transport_layer/context.hpp]:
```suggestion
                communicator_type get_communicator()
```

File: include/ghex/transport_layer/util/barrier.hpp (hunk 13 above)
Comment (mbianco) [inline include/ghex/transport_layer/util/barrier.hpp:108]:
```suggestion
                 * This function synchronize the threads in a rank. The number of threads that need to participate
                 * is indicated in the construction of the barrier object, whose reference is shared among the
                 * participating threads.
```

File: tests/transport/test_ucx_context.cpp [hunk 14]
```diff
@@ -22,48 +21,46 @@ namespace ghex = gridtools::ghex;
...
         {
-            payload[0] = token.id();
+            payload[0] = id;
             payload[1] = comm.rank();
-            comm.send(payload, 0, token.id()).wait();
+            comm.send(payload, 0, id).wait();
         }
 
-        comm.barrier();
+        //comm.barrier();
```
Comment (mbianco) [inline tests/transport/test_ucx_context.cpp:57]:
```suggestion
```
Doe not seem to be necessary

File: include/ghex/transport_layer/ucx/request.hpp (hunk 6 above)
Comment (mbianco) [reply include/ghex/transport_layer/ucx/request.hpp:104]: I re-introduced the alignment (hoping that this would fix the bug, but it did not)

File: benchmarks/transport/ghex_p2p_bi_cb_avail_mt.cpp [hunk 15]
```diff
@@ -151,7 +152,13 @@ int main(int argc, char *argv[])
                 make_zero(rmsgs[j]);
             }
 
-            comm.barrier();
+#ifdef USE_OPENMP
+#pragma omp single
+#endif
+            barrier.rank_barrier(comm);
```
Comment (boeschf) [inline benchmarks/transport/ghex_p2p_bi_cb_avail_mt.cpp:158]: why do we need the `pragma omp single` here? Can a general barrier work here?

File: include/ghex/transport_layer/context.hpp [hunk 16]
```diff
@@ -88,13 +84,6 @@ namespace gridtools {
...
-                /** @brief return a reference to the thread-shared thread primitives instance.
-                  * This function is thread-safe. */
-                thread_primitives_type& thread_primitives() noexcept
-                {
-                    return m_thread_primitives;
-                }
-
                 /** @brief return a special per-rank setup communicator.
                   * This function is not thread-safe and should only be used in the serial part of the code. */
                 mpi::setup_communicator get_setup_communicator()
```
Comment (boeschf) [inline include/ghex/transport_layer/context.hpp:89]: maybe we don't need this function anymore at all.

File: include/ghex/transport_layer/ucx/communicator.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (boeschf) [reply include/ghex/transport_layer/ucx/communicator.hpp:18]: ok, no objections

File: include/ghex/transport_layer/util/barrier.hpp (hunk 12 above)
Comment (boeschf) [reply include/ghex/transport_layer/util/barrier.hpp:143]: @angainor I think `ex==1` is correct here. same goes for `in_node1`. @mbianco can you comment?

**Review by boeschf — COMMENTED** [review]
please address the remaining comments/ open questions

File: include/ghex/transport_layer/mpi/context.hpp [hunk 17]
```diff
@@ -19,59 +20,49 @@ namespace gridtools {
...
                 {
-                    if (!m_states[t.id()])
-                    {
-                        m_tokens[t.id()] = t;
-                        m_states[t.id()] = std::make_unique<state_type>(&m_tokens[t.id()]);
-                    }
-                    return {&m_shared_state, m_states[t.id()].get()};
+                    std::lock_guard<std::mutex> lock(m_mutex); // we need to guard only the isertion in the vector, but this is not a performance critical section
+                    m_states.push_back(std::make_unique<state_type>());
+                    return {&m_shared_state, m_states[m_states.size()-1].get()};
```
[the remark this answers is not in the document]
Comment (mbianco) [reply include/ghex/transport_layer/mpi/context.hpp:55]: yes, that's the point.

File: tests/transport/test_send_recv.cpp [hunk 18]
```diff
@@ -58,7 +56,7 @@ auto test_ring_send_recv_ft(CommType& comm, std::size_t buffer_size)
     data_ptr = reinterpret_cast<int*>(smsg.data());
     *data_ptr = rank;
 
-    comm.barrier();
+    MPI_Barrier(MPI_COMM_WORLD);
```
[the remark this answers is not in the document]
Comment (mbianco) [reply tests/transport/test_send_recv.cpp:59]: AFAIK this is used just to synchronize the threads for timig

File: include/ghex/transport_layer/mpi/context.hpp [hunk 19]
```diff
@@ -19,59 +20,49 @@ namespace gridtools {
...
-                transport_context(ThreadPrimitives& tp, MPI_Comm mpi_comm, Args&&...)
-                : m_thread_primitives(tp)
-                , m_comm{mpi_comm}
-                , m_tokens(tp.size())
-                , m_shared_state(mpi_comm, this, &tp)
-                , m_state(nullptr)
-                , m_states(tp.size())
+                transport_context(MPI_Comm mpi_comm, Args&&...)
+                : m_comm{mpi_comm}
+                , m_shared_state(mpi_comm, this)
```
[the remark this answers is not in the document]
Comment (mbianco) [reply include/ghex/transport_layer/mpi/context.hpp:41]: it is default initialized, as m_mutex

File: include/ghex/transport_layer/ucx/worker.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (mbianco) [reply include/ghex/transport_layer/ucx/worker.hpp:88]: done

File: include/ghex/pattern.hpp [hunk 20]
```diff
@@ -116,8 +115,8 @@ namespace gridtools {
          * @param d_range range of local domains
          * @return iterable of patterns (one per domain)
          */
-        template<typename GridType, typename Transport, typename ThreadPrimitives, typename HaloGenerator, typename RecvDomainIdsGen, typename DomainRange>
-        auto make_pattern(tl::context<Transport,ThreadPrimitives>& context, HaloGenerator&& hgen, RecvDomainIdsGen&& recv_domain_ids_gen, DomainRange&& d_range)
+        template<typename GridType, typename Transport, typename HaloGenerator, typename RecvDomainIdsGen, typename DomainRange>
```
[the remark this answers is not in the document]
Comment (boeschf) [reply include/ghex/pattern.hpp:139]: this is still there

File: include/ghex/transport_layer/context.hpp [hunk 21]
```diff
@@ -114,16 +103,9 @@ namespace gridtools {
...
-                    return m_transport_context.get_communicator(t);
-                }
-
-                /** @brief return a per-thread thread token.
-                  * This function is thread-safe. */
-                thread_token get_token() noexcept
+                communicator_type get_communicator()
                 {
-                    return m_thread_primitives.get_token();
+                    return m_transport_context.get_communicator(/*t*/);
```
Comment (boeschf) [inline include/ghex/transport_layer/context.hpp:108]:
```suggestion
                    return m_transport_context.get_communicator();
```

File: include/ghex/transport_layer/ucx/communicator.hpp (hunk 3 above)
Comment (boeschf) [reply include/ghex/transport_layer/ucx/communicator.hpp:122]: shouldn't we use the mutex from the recv worker? since the send worker is not shared?
also shouldn't there be a `std::remove_reference_t` in the `decltype`?

File: include/ghex/transport_layer/ucx/communicator.hpp [hunk 22]
```diff
@@ -163,25 +160,23 @@ namespace gridtools {
...
-                            {
-                                p+= ucp_worker_progress(m_ucp_rw);
-                                p+= ucp_worker_progress(m_ucp_rw);
-                                status.m_num_recvs = std::exchange(m_recv_worker->m_progressed_recvs, 0);
-                                status.m_num_cancels = std::exchange(m_recv_worker->m_progressed_cancels, 0);
-                            }
-                        );
-                        //return p;
+                        {
+                        std::lock_guard<decltype(m_send_worker->mutex())> lock(m_send_worker->mutex());
```
Comment (boeschf) [inline include/ghex/transport_layer/ucx/communicator.hpp:169]: here again for clarity: use the mutex from context directly (or from recv worker). remove_reference_t?

**Review by boeschf — COMMENTED** [review]
almost ready. would be good if you could quickly add the clarifications w.r.t. the mutex.

File: benchmarks/transport/ghex_p2p_bi_cb_avail_mt.cpp (hunk 15 above)
Comment (mbianco) [reply benchmarks/transport/ghex_p2p_bi_cb_avail_mt.cpp:158]: I tried reproduce the behavior of previous implementation, so Iwould keep the omp single here

File: tests/communication_object.cpp [hunk 23]
```diff
@@ -101,7 +100,7 @@ TEST(communication_object, constructor) {
     using coordinate_t = domain_descriptor_t::coordinate_type;
     using halo_generator_t = domain_descriptor_t::halo_generator_type; //gridtools::structured::halo_generator<domain_id_t, 3>;
 
-    auto context_ptr = gridtools::ghex::tl::context_factory<transport,threading>::create(1, MPI_COMM_WORLD);
+    auto context_ptr = gridtools::ghex::tl::context_factory<transport>::create(1, MPI_COMM_WORLD);
```
[the remark this answers is not in the document]
Comment (mbianco) [reply tests/communication_object.cpp:103]: The fact that the tests compile tells me this file is unused...

File: include/ghex/transport_layer/ucx/communicator.hpp (hunk 22 above)
Comment (mbianco) [reply include/ghex/transport_layer/ucx/communicator.hpp:169]: I do not think remove_reference helps here, since this is an internal function and the fact that it works without adding facilities tells me we do not need it. Adding it, would mean, to me, that I do not understand what's going on (which may be true, btw). We decided to not make this change, since there are some issues with forward declarations otherwise, which would make the code harder to read.

File: include/ghex/transport_layer/mpi/context.hpp [hunk 24]
```diff
@@ -19,68 +20,64 @@ namespace gridtools {
...
-                , m_rank_topology{t}
-                , m_comm{mpi_comm}
-                , m_tokens(tp.size())
-                , m_shared_state(mpi_comm, this, &tp)
-                , m_state(nullptr)
-                , m_states(tp.size())
+                transport_context(const mpi::rank_topology& t, MPI_Comm mpi_comm, Args&&...)
+                    : m_rank_topology{t}
+                    , m_comm{mpi_comm}
+                    , m_shared_state(mpi_comm, this)
```
[the remark this answers is not in the document]
Comment (mbianco) [reply include/ghex/transport_layer/mpi/context.hpp:43]: not it is default initialized, as it should be (before it had primitives, but now they are not there anymore, so it should be fine.

[the remark this answers is not in the document]
Comment (mbianco) [reply include/ghex/transport_layer/mpi/context.hpp:43]: mark as resolved if you are satisfied @bettiolm

File: tests/communication_object.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (bettiolm) [reply tests/communication_object.cpp]: Yep, that test is no longer on the list

File: include/ghex/transport_layer/mpi/context.hpp (hunk 24 above)
[the remark this answers is not in the document]
Comment (bettiolm) [reply include/ghex/transport_layer/mpi/context.hpp:43]: Yep

**Review by boeschf — APPROVED** [review]

**Review by bettiolm — APPROVED** [review]
